### PR TITLE
Add PHS3302 D0 lifetime assignment 

### PR DESCRIPTION
**Pull request by TomHadavizadeh**
- Added notebook/python scripts for D0 lifetime assignment 
- Included a new dependency (uproot4)

### CODE REVIEW

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+## Lets try plotting the momentum of the kaon and pion
+fig, ax = plt.subplots(2, 3,figsize=(10,5))
+ax[0,0].hist(K_PX,bins=100,label="K_PX")
+ax[0,0].set_xlabel('$p_{x}(K^-)$ [MeV]')
+ax[0,0].set_ylabel('Candidates')
+
+ax[0,1].hist(K_PY,bins=100)
+ax[0,1].set_xlabel('$p_{y}(K^-)$ [MeV]')
+ax[0,1].set_ylabel('Candidates')
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:81]: I find px and py plots a bit hard to comprehend Would it be better to look at Pt?

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+# 
+# ### Part 2a: $D^0$ mass
+# The data contains the 3-momenta of the kaon and pion:
+# $$
+# \vec{\mathbf{p}} = (p_{x},p_{y},p_{z})
+# $$
+# We need the 4-momenta, which means adding the energy of the particles as well. At the LHCb experiment we don't measure the particle energy directly. Instead we use particle identification detectors to determine what type of particle the various the long-lived particles are (e.g. kaons or pions) and use the externally measured values of their masses and our measured momentums to compute the energy using 
+# $$
+# E^2 = |\vec{\mathbf{p}}|^2 + m^2
+# $$
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:146]: This they should just know.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+# 
+# The mass can be calculated from the sum of the momentum 4-vectors for the kaon and pion 
+# $$
+# \mathbf{p}_{D^{0}} = \mathbf{p}_{K} + \mathbf{p}_{\pi},
+# $$
+# 
+# In the rest frame of the $D^0$ meson we will have $\mathbf{p}_{D^0} = \left[m_{D^0},0,0,0\right]$
+# $$
+# m_{D^{0}}^2 = \mathbf{p}_{D^{0}}.\mathbf{p}_{D^{0}} = (\mathbf{p}_{K} + \mathbf{p}_{\pi})^2.
+# $$
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:179]: This they should just know as well.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+# **Task 2:** *[2 marks]* Calculate the $D^0$ mass using the all of the components of the kaon and pion 4-momentum. 
+
+
+
+## D0_M = 
+
+
+# A loose selection has already been applied to the sample we are analysing. This means that the signal contains both real $D^0\to K \pi$ decays as well as backgrounds from combinations of a kaon and pion that didn't originate from a $D^0$ meson.
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:187]: You start by introducing the unknown word selection, and then explain it afterwards. Maybe turn around.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+# \Delta\vec{\mathbf{x}} = \vec{\mathbf{v}}t
+# $$
+# 
+# We must account for special relativity when trying to determine the lifetime of the $D^0$ meson. We can rewrite this equation using relations between the momentum and velocity, and between lab frame time and proper time:
+# 
+# $$
+# \vec{\mathbf{p}} = \gamma m \vec{\mathbf{v}}\\
+# t = \gamma \tau.
+# $$
+#
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:227]: If you just remind them of this, and encourage them to carry the 3-vectors of p and v forward, they should be able to get the expression themselves.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+# $$
+# 
+# If we use $\vec{\mathbf{p}}$ in units of $[{\rm MeV}]$, $m$ in units of $[{\rm MeV}]$ and $\Delta\vec{\mathbf{x}}$ in units of $[\rm mm]$, the proper time $\tau$ will be calculated in units of $[\rm mm]$ (i.e. it's actually a length. This is because we are using natural units for the momentum and mass but SI units for the positions). 
+# 
+# We can convert $[\rm mm]$ into $[\rm ps]$ by dividing  by the speed of light in units of $[\rm mm \,ps^{-1}]$ (given at the top of the script).
+
+# *Aside:* this method is of determining the lifetime is only approxmate. This is because the particles travel through a magnetic field which bends their trajectory, allowing us to measure their momentum. As a result, the momentum 3-vectors of the kaons and pions point in a slightly different direction depending on where we are measuring them. However, the momenta we used to find the vertex assumed the kaon and pion originated at the proton-proton interaction, which we don't expect to be true! Therefore finding the best estimate of the decay time requires iteratively changing the decay position and recalculating the momentum until the values converge.
```
Comment (egede) [inline monashspa/PHS3302/d0_lifetime/d0_lifetime.py:241]: Maybe just delete this. As we do not use it, it wil just cause confusion.

**Review by egede — CHANGES_REQUESTED** [review]
This looks really good overall. I have just made a few comments, mainly that there are some places where they can derive the expressions themselves.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 1 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:81]: Good suggestion! That will also allow me to include an example of using numpy to calculate new quantities.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 2 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:146]: Thanks, I've removed this

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 3 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:179]: I've taken this out too

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 4 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:187]: Thanks, I've reworded this bit

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 5 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:227]: Good idea, I've created a new task asking them to do the last two steps themselves.

File: monashspa/PHS3302/d0_lifetime/d0_lifetime.py (hunk 6 above)
Comment (TomHadavizadeh) [reply monashspa/PHS3302/d0_lifetime/d0_lifetime.py:241]: I've taken out the aside

**Review by egede — APPROVED** [review]
Looks good now. Just remove the WIP tag when ready and I will merge.

**Pull request merged**